src/query/functions/src/scalars/map.rs

```diff
@@ -0,0 +1,129 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::hash::Hash;
+use std::sync::Arc;
+
+use common_expression::types::array::ArrayColumnBuilder;
+use common_expression::types::map::KvPair;
+use common_expression::types::ArrayType;
+use common_expression::types::DataType;
+use common_expression::types::GenericType;
+use common_expression::types::ValueType;
+use common_expression::Column;
+use common_expression::Function;
+use common_expression::FunctionDomain;
+use common_expression::FunctionProperty;
+use common_expression::FunctionRegistry;
+use common_expression::FunctionSignature;
+use common_expression::Scalar;
+use common_expression::Value;
+use common_expression::ValueRef;
+use common_hashtable::StackHashSet;
+use siphasher::sip128::Hasher128;
+use siphasher::sip128::SipHasher24;
+
+pub fn register(registry: &mut FunctionRegistry) {
+    registry.register_function_factory("map", |_, args_type| {
+        if args_type.len() != 2 {
+            return None;
+        }
```

Comment:
it's better to use `registry.register_passthrough_nullable::<ArrayType<GenericType<0>>, ArrayType<GenericType<1>>, MapType<GenericType<0>, GenericType<1>>>` and then check the key type via `ctx.generics[0]`